import os
import sys
from PySide6.QtWidgets import QMainWindow, QListWidgetItem
from PySide6.QtUiTools import QUiLoader
from PySide6.QtCore import QFile, Qt

class VoiceInterface(QMainWindow):
    def __init__(self, client_logic):
        super().__init__()
        self.client = client_logic
        
        # 1. –ó–∞–≥—Ä—É–∑–∫–∞ UI
        loader = QUiLoader()
        script_dir = os.path.dirname(os.path.abspath(__file__))
        ui_path = os.path.join(script_dir, "interface", "main_interface.ui")
        
        ui_file = QFile(ui_path)
        if not ui_file.open(QFile.ReadOnly):
            print(f"–û–®–ò–ë–ö–ê: –ù–µ —É–¥–∞–ª–æ—Å—å –æ—Ç–∫—Ä—ã—Ç—å {ui_path}")
            return

        self.ui = loader.load(ui_file)
        ui_file.close()

        if self.ui:
            self.setCentralWidget(self.ui.centralwidget)
            self.setWindowTitle(self.ui.windowTitle() or "Voice Chat v1.0")
            self.setFixedSize(self.ui.size()) 
            
            # –ü—Ä–∏–≤—è–∑—ã–≤–∞–µ–º –ª–æ–≥–∏–∫—É –∫–ª–∏–µ–Ω—Ç–∞ –∫ –º–µ—Ç–æ–¥–∞–º –∏–Ω—Ç–µ—Ä—Ñ–µ–π—Å–∞ (Callback)
            self.client.on_users_received = self.handle_users_update
            self.client.on_messages_received = self.handle_messages_update
            self.client.on_server_name_received = self.handle_server_name
            
            self.init_widgets()
            self.update_server_list_ui()

    def init_widgets(self):
        """–ù–∞—Å—Ç—Ä–æ–π–∫–∞ –Ω–∞—á–∞–ª—å–Ω—ã—Ö —Å–æ—Å—Ç–æ—è–Ω–∏–π –≤–∏–¥–∂–µ—Ç–æ–≤ –∏ –∫–æ–Ω–Ω–µ–∫—Ç–æ–≤"""
        # –°–ª–∞–π–¥–µ—Ä—ã
        self.ui.SoundVolume.setRange(0, 100)
        self.ui.MicVolume.setRange(0, 100)
        self.ui.ThresholdVolume.setRange(0, 1000)
        
        self.ui.SoundVolume.setValue(self.client.sound_volume)
        self.ui.MicVolume.setValue(self.client.mic_volume)
        self.ui.ThresholdVolume.setValue(self.client.threshold)

        # –ö–Ω–æ–ø–∫–∏
        self.ui.AddIP.clicked.connect(self.add_server_to_list)
        
        # –ò–∑–º–µ–Ω–µ–Ω–æ: —Ä–µ–∞–≥–∏—Ä—É–µ–º –Ω–∞ —Å–º–µ–Ω—É —Ç–µ–∫—É—â–µ–≥–æ —ç–ª–µ–º–µ–Ω—Ç–∞ (–ª–µ–≤—ã–π –∫–ª–∏–∫/–≤—ã–±–æ—Ä –≤ —Å–ø–∏—Å–∫–µ)
        self.ui.ServerList.itemClicked.connect(self.connect_server)
        
        self.ui.MicOnOff.clicked.connect(self.toggle_mic)
        self.ui.SoundOnOff.clicked.connect(self.toggle_sound)
        
        if hasattr(self.ui, 'SendMessage'):
            self.ui.SendMessage.clicked.connect(self.send_text_msg)
        
        # –û–±–Ω–æ–≤–ª–µ–Ω–∏–µ –ø–∞—Ä–∞–º–µ—Ç—Ä–æ–≤
        self.ui.SoundVolume.valueChanged.connect(self.sync_settings)
        self.ui.MicVolume.valueChanged.connect(self.sync_settings)
        self.ui.ThresholdVolume.valueChanged.connect(self.sync_settings)

    # --- –û–ë–†–ê–ë–û–¢–ß–ò–ö–ò API (Incoming Data) ---

    def handle_users_update(self, users):
        self.ui.UserList.clear()
        for user in users:
            item = QListWidgetItem(f"üë§ {user['name']}")
            self.ui.UserList.addItem(item)

    def handle_messages_update(self, messages):
        for msg in messages:
            text = f"{msg['id']}: {msg['message']}" if 'id' in msg else msg['message']
            self.ui.MessageList.addItem(text)
        self.ui.MessageList.scrollToBottom()

    def handle_server_name(self, name):
        """–ö–æ–≥–¥–∞ —Å–µ—Ä–≤–µ—Ä –ø—Ä–∏—Å–ª–∞–ª —Å–≤–æ–µ –∏–º—è, –æ–±–Ω–æ–≤–ª—è–µ–º —Å—Ç–∞—Ç—É—Å–±–∞—Ä"""
        if hasattr(self.ui, 'statusbar'):
            self.ui.statusbar.showMessage(f"–ü–æ–¥–∫–ª—é—á–µ–Ω–æ –∫: {name}")

    # --- –õ–û–ì–ò–ö–ê –ò–ù–¢–ï–†–§–ï–ô–°–ê (Outgoing Actions) ---

    def add_server_to_list(self):
        """–î–æ–±–∞–≤–ª—è–µ—Ç IP –≤ —Å–ø–∏—Å–æ–∫ —Å–µ—Ä–≤–µ—Ä–æ–≤"""
        ip = self.ui.EnterIP.toPlainText().strip()
        if ip:
            # –ü—Ä–∏ –¥–æ–±–∞–≤–ª–µ–Ω–∏–∏ –∏—Å–ø–æ–ª—å–∑—É–µ–º IP –∫–∞–∫ –≤—Ä–µ–º–µ–Ω–Ω–æ–µ –∏–º—è
            new_srv = {"ip": ip, "port": "5555", "name": ip}
            self.client.server_list.append(new_srv)
            self.client.save_servers()
            
            self.update_server_list_ui()
            self.ui.EnterIP.clear()
            if hasattr(self.ui, 'statusbar'):
                self.ui.statusbar.showMessage(f"IP {ip} –¥–æ–±–∞–≤–ª–µ–Ω –≤ —Å–ø–∏—Å–æ–∫", 2000)

    def connect_server(self, item):
        """–°—Ä–∞–±–∞—Ç—ã–≤–∞–µ—Ç –ø—Ä–∏ –∫–ª–∏–∫–µ –Ω–∞ —Å–µ—Ä–≤–µ—Ä –≤ —Å–ø–∏—Å–∫–µ"""
        srv = item.data(Qt.UserRole)
        
        # –ë–µ—Ä–µ–º –Ω–∏–∫ –ø–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª—è –∏–∑ –ø–æ–ª—è EnterName
        user_nickname = self.ui.EnterName.toPlainText().strip() or "User"
        
        self.client.stop() 
        self.client.connect_to_server(srv['ip'], int(srv['port']), user_name=user_nickname)
        
        # –í–∏–∑—É–∞–ª—å–Ω–∞—è –æ—á–∏—Å—Ç–∫–∞ –ø—Ä–∏ –ø–µ—Ä–µ–∫–ª—é—á–µ–Ω–∏–∏
        self.ui.MessageList.clear()
        self.ui.UserList.clear()
        if hasattr(self.ui, 'statusbar'):
            self.ui.statusbar.showMessage(f"–ü–æ–ø—ã—Ç–∫–∞ –ø–æ–¥–∫–ª—é—á–µ–Ω–∏—è –∫ {srv['ip']}...")

    def send_text_msg(self):
        text = self.ui.EnterMessage.toPlainText().strip()
        if text:
            self.client.send_text_message(text)
            self.ui.EnterMessage.clear()

    def sync_settings(self):
        self.client.sound_volume = self.ui.SoundVolume.value()
        self.client.mic_volume = self.ui.MicVolume.value()
        self.client.threshold = self.ui.ThresholdVolume.value()

    def update_server_list_ui(self):
        self.ui.ServerList.clear()
        for srv in self.client.server_list:
            display_name = f"üåê {srv['ip']}"
            item = QListWidgetItem(display_name)
            item.setData(Qt.UserRole, srv)
            self.ui.ServerList.addItem(item)

    def toggle_mic(self):
        self.client.is_mic_muted = not self.client.is_mic_muted
        style = "background-color: #ff4444; color: white; border-radius: 5px;" if self.client.is_mic_muted else ""
        self.ui.MicOnOff.setStyleSheet(style)

    def toggle_sound(self):
        self.client.is_sound_muted = not self.client.is_sound_muted
        style = "background-color: #ff4444; color: white; border-radius: 5px;" if self.client.is_sound_muted else ""
        self.ui.SoundOnOff.setStyleSheet(style)